### PR TITLE
Pypsa substations

### PR DESCRIPTION
**Pull request by FabianHofmann**
Introduce a functionality to import substations from arbitrary pypsa networks. The code is mostly aligned to the convention in pypsa-eur(-sec)

### Purpose
What is the larger goal of this change?

### What the code is doing
How is the purpose executed?

### Testing
How did you test this change (unit/functional testing, manual testing, etc.)? 

### Where to look
* It's helpful to clarify where your new code lives if you moved files around or there could be confusion/

* What files are most important?

### Usage Example/Visuals
How the code can be used and/or images of any graphs, tables or other visuals (not always applicable). 

### Time estimate
How long will it take for reviewers and observers to understand this code change?

### CODE REVIEW

**Comment by rouille** [comment]
Can you rebase onto `ben/import`, it has been rebased onto `develop`

**Comment by rouille** [comment]
@FabianHofmann, is that ready to be reviewed?

**Comment by rouille** [comment]
I did the following:
```
>>> from powersimdata.input.converter.pypsa_to_grid import FromPyPSA
>>> from powersimdata.network.europe_tub.model import TUB
>>> tub = TUB("Europe", reduction=256)
Title: PyPSA-Eur: An Open Optimisation Model of the European Transmission System (Dataset)
Keywords: power system model, capacity expansion model, energy system model, electricity system model, python
Publication date: 2022-09-10
DOI: 10.5281/zenodo.6913032
Total size: 2075.6 MB

Link: https://zenodo.org/api/files/7528030f-2993-4f54-83e4-9fdd223b9ba5/networks.zip   size: 2075.6 MB
100% [....................................................................] 2176404814 / 2176404814	
Checksum is correct. (d735594db370f8cd867a79a58ab5841d)
All files have been downloaded.
>>> tub.build()
INFO:pypsa.io:Imported network elec_s_256_ec.nc has buses, carriers, generators, lines, links, loads, storage_units, stores
>>> pypsa_to_grid = FromPyPSA(tub.network)
>>> pypsa_to_grid.sub
         name  interconnect_sub_id        lat        lon interconnect    pypsa_y    pypsa_x
sub_id                                                                                     
AL1 0   AL1 0                  NaN  41.587568  19.897914           AL  41.587568  19.897914
AT1 0   AT1 0                  NaN  47.232051   9.867096           AT  47.232051   9.867096
AT1 1   AT1 1                  NaN  47.516333  14.286742           AT  47.516333  14.286742
AT1 2   AT1 2                  NaN  47.548309  12.816925           AT  47.548309  12.816925
AT1 3   AT1 3                  NaN  47.889033  16.198929           AT  47.889033  16.198929
...       ...                  ...        ...        ...          ...        ...        ...
SE2 8   SE2 8                  NaN  63.347806  13.221588           SE  63.347806  13.221588
SE2 9   SE2 9                  NaN  62.574400  15.212250           SE  62.574400  15.212250
SI1 0   SI1 0                  NaN  46.071068  14.693952           SI  46.071068  14.693952
SK1 0   SK1 0                  NaN  48.872723  20.059073           SK  48.872723  20.059073
SK1 1   SK1 1                  NaN  48.417068  17.917381           SK  48.417068  17.917381

[256 rows x 7 columns]
```
It looks like `interconnect_sub_id` are all NaNs. @BainanXia, do you remember what this field is for? If I do:
```
>>> from powersimdata import Grid
>>> grid = Grid("USA")
Reading bus.csv
Reading plant.csv
Reading gencost.csv
Reading branch.csv
Reading dcline.csv
Reading sub.csv
Reading bus2sub.csv
Reading zone.csv
>>> grid.sub
                           name  interconnect_sub_id        lat         lon interconnect
sub_id                                                                                  
0                      AUBURN 1                    1  43.976100  -70.221100      Eastern
1                      AUBURN 2                    2  44.101000  -70.305300      Eastern
2                      AUBURN 3                    3  44.179900  -70.186000      Eastern
3                      AUBURN 4                    4  44.143300  -70.224100      Eastern
4                      DURHAM 1                    5  44.020700  -70.180500      Eastern
...                         ...                  ...        ...         ...          ...
41078           OREGON PORTLAND                81019  45.435890 -124.324466      Western
41079        WASHINGTON RAYMOND                81020  46.576647 -124.349185      Western
41080   WASHINGTON OCEAN SHORES                81021  46.979165 -124.478275      Western
41081         WASHINGTON QUEETS                81022  47.534671 -124.655429      Western
41082         WASHINGTON OZETTE                81023  48.168686 -125.050937      Western

[41083 rows x 5 columns]
>>> import matplotlib.pyplot as plt
>>> grid.sub["interconnect_sub_id"].plot()
<AxesSubplot:xlabel='sub_id'>
>>> plt.show()
``` 
interconnect_sub_id

The relationship is not straightforward. It is linear for most of Eastern, linear for Texas and non-linear for Wesern.

**Comment by BainanXia** [comment]
@rouille I think that's the legacy sub id from the original aux files. Recall the first version of the US grid was concatenated by three TAMU grids and the sub ids weren't unique. I don't know why the Western ones are not linear by themselves though. I think it is okay to leave them NAN given they are not used or checked elsewhere?

**Comment by rouille** [comment]
@FabianHofmann. It seems that `bus2sub` is not properly populated:
```
>>> pypsa_to_grid.bus2sub
               sub_id interconnect
bus_id                            
AL1 0             NaN           AL
AT1 0             NaN           AT
AT1 1             NaN           AT
AT1 2             NaN           AT
AT1 3             NaN           AT
...               ...          ...
SE2 8 battery     NaN           SE
SE2 9 battery     NaN           SE
SI1 0 battery     NaN           SI
SK1 0 battery     NaN           SK
SK1 1 battery     NaN           SK

[768 rows x 2 columns]
```
I would expect the the `sub_id` columns to be the indices of the `sub` data frame for pypsa-eur since you have one bus per substations.

File: powersimdata/input/converter/pypsa_to_grid.py [hunk 1]
```diff
@@ -230,14 +224,33 @@ def _read_network(self, n, add_pypsa_cols=True):
...
+            sub["interconnect"] = sub.index.str.extract(interconnect_pattern).values
+            sub["name"] = sub.index
+            sub_pypsa = bus_pypsa[bus_pypsa.index.str.match(sub_pattern)][
+                sub_pypsa_cols
+            ]
+
+            sub_pattern = "([A-Z][A-Z]\d+\s\d+)$"
+            bus2sub = pd.DataFrame(
+                {
+                    "name": bus.index.str.extract(sub_pattern)[0].values,
```
Comment (rouille) [inline powersimdata/input/converter/pypsa_to_grid.py:242]: Should the column name be "sub_id" instead of "name"?

**Comment by rouille** [comment]
The `sub` and `bus2sub` data frames have a non-NaN value for "interconnect" (country abbreviation in the comment above where pypsa-eur is used as an example) while the other data frames of the `Grid` object have values in the "interconnect" columns set to NaN:
```
>>> pypsa_to_grid.plant["interconnect"]
plant_id
AL1 0 offwind-ac   NaN
AL1 0 oil          NaN
AL1 0 onwind       NaN
AL1 0 ror          NaN
AL1 0 solar        NaN
                    ..
SK1 1 lignite      NaN
SK1 1 nuclear      NaN
SK1 1 onwind       NaN
SK1 1 ror          NaN
SK1 1 solar        NaN
Name: interconnect, Length: 1325, dtype: float64
>>> pypsa_to_grid.bus["interconnect"]
bus_id
AL1 0           NaN
AT1 0           NaN
AT1 1           NaN
AT1 2           NaN
AT1 3           NaN
                 ..
SE2 8 battery   NaN
SE2 9 battery   NaN
SI1 0 battery   NaN
SK1 0 battery   NaN
SK1 1 battery   NaN
Name: interconnect, Length: 768, dtype: float64
>>> pypsa_to_grid.dcline["from_interconnect"]
dcline_id
14811                      NaN
5640                       NaN
13589                      NaN
14802                      NaN
14825                      NaN
                            ..
SE2 8 battery discharger   NaN
SE2 9 battery discharger   NaN
SI1 0 battery discharger   NaN
SK1 0 battery discharger   NaN
SK1 1 battery discharger   NaN
Name: from_interconnect, Length: 1065, dtype: float64
```
My recommendation would be to let the interconnect values set to NaN in all data frames and let the model immutables fill out the information as we will do for `zone2id`:
```
>>> pypsa_to_grid.zone2id
{}
```
What do you think @BainanXia and @jenhagg?

File: powersimdata/input/converter/pypsa_to_grid.py (hunk 1 above)
Comment (FabianHofmann) [reply powersimdata/input/converter/pypsa_to_grid.py]: Yes, that was the problem thanks!

**Comment by FabianHofmann** [comment]
@rouille I removed the interconnect values as you suggested.

**Comment by rouille** [comment]
@FabianHofmann, I noticed that the (lat, lon) are not propagated to all data frames. For instance (lat, lon) and (from_lat, from_lon) are set to NaN for the `plant` and `branch` data frames, respectively. There might be other cases.

**Comment by FabianHofmann** [comment]
@rouille this is right. Should we however tackle this is another PR? (This seems not to be related to the code changes I made)

**Comment by rouille** [comment]
> @rouille this is right. Should we however tackle this is another PR? (This seems not to be related to the code changes I made)

Sounds good

**Comment by rouille** [comment]
After your latest commit, the `sub_id` is correctly set in the `bus2sub` data frame and the `name` are NaN in the `sub` data frame:
```
>>> pypsa_to_grid.sub
        name  interconnect_sub_id        lat        lon  interconnect    pypsa_y    pypsa_x
sub_id                                                                                     
AL1 0    NaN                  NaN  41.587568  19.897914           NaN  41.587568  19.897914
AT1 0    NaN                  NaN  47.232051   9.867096           NaN  47.232051   9.867096
AT1 1    NaN                  NaN  47.516333  14.286742           NaN  47.516333  14.286742
AT1 2    NaN                  NaN  47.548309  12.816925           NaN  47.548309  12.816925
AT1 3    NaN                  NaN  47.889033  16.198929           NaN  47.889033  16.198929
...      ...                  ...        ...        ...           ...        ...        ...
SE2 8    NaN                  NaN  63.347806  13.221588           NaN  63.347806  13.221588
SE2 9    NaN                  NaN  62.574400  15.212250           NaN  62.574400  15.212250
SI1 0    NaN                  NaN  46.071068  14.693952           NaN  46.071068  14.693952
SK1 0    NaN                  NaN  48.872723  20.059073           NaN  48.872723  20.059073
SK1 1    NaN                  NaN  48.417068  17.917381           NaN  48.417068  17.917381

[256 rows x 7 columns]
>>> pypsa_to_grid.bus2sub
              sub_id  interconnect
bus_id                            
AL1 0          AL1 0           NaN
AT1 0          AT1 0           NaN
AT1 1          AT1 1           NaN
AT1 2          AT1 2           NaN
AT1 3          AT1 3           NaN
...              ...           ...
SE2 8 battery  SE2 8           NaN
SE2 9 battery  SE2 9           NaN
SI1 0 battery  SI1 0           NaN
SK1 0 battery  SK1 0           NaN
SK1 1 battery  SK1 1           NaN

[768 rows x 2 columns]
>>> 
```
@BainanXia, are we fine with `name` being NaN? Prior to the last coomit it was set to `sub_id`, see [comment above](https://github.com/Breakthrough-Energy/PowerSimData/pull/674#issuecomment-1281879406).

**Comment by BainanXia** [comment]
@rouille I think it should be okay. Optimization wise, it won't cause any problem for sure. We are not using the substation names elsewhere either.

**Review by rouille — APPROVED** [review]
Thanks

**Pull request merged**